x-pack/legacy/plugins/ml/public/timeseriesexplorer/timeseriesexplorer_utils/get_focus_data.ts

```diff
@@ -0,0 +1,169 @@
+}
+
+export interface FocusData {
+  focusChartData: any;
+  anomalyRecords: any;
+  scheduledEvents: any;
+  showForecastCheckbox: any;
+  focusAnnotationData: any;
+  focusForecastData: any;
+}
+
+export function getFocusData(
+  criteriaFields: CriteriaField[],
+  detectorIndex: number,
+  focusAggregationInterval: Interval,
+  forecastId: string,
+  modelPlotEnabled: boolean,
+  nonBlankEntities: any[],
+  searchBounds: any,
+  selectedJob: Job
+): Observable<FocusData> {
+  return forkJoin([
+    // Query 1 - load metric data across selected time range.
+    mlTimeSeriesSearchService.getMetricData(
+      selectedJob,
+      detectorIndex,
+      nonBlankEntities,
+      searchBounds.min.valueOf(),
+      searchBounds.max.valueOf(),
+      focusAggregationInterval.expression
+    ),
+    // Query 2 - load all the records across selected time range for the chart anomaly markers.
+    mlResultsService.getRecordsForCriteria(
+      [selectedJob.job_id],
+      criteriaFields,
+      0,
+      searchBounds.min.valueOf(),
+      searchBounds.max.valueOf(),
+      ANOMALIES_TABLE_DEFAULT_QUERY_SIZE
+    ),
+    // Query 3 - load any scheduled events for the selected job.
+    mlResultsService.getScheduledEventsByBucket(
+      [selectedJob.job_id],
+      searchBounds.min.valueOf(),
+      searchBounds.max.valueOf(),
+      focusAggregationInterval.expression,
+      1,
+      MAX_SCHEDULED_EVENTS
+    ),
+    // Query 4 - load any annotations for the selected job.
+    mlAnnotationsEnabled
+      ? ml.annotations
+          .getAnnotations({
+            jobIds: [selectedJob.job_id],
+            earliestMs: searchBounds.min.valueOf(),
+            latestMs: searchBounds.max.valueOf(),
+            maxAnnotations: ANNOTATIONS_TABLE_DEFAULT_QUERY_SIZE,
+          })
+          .pipe(
+            catchError(() => {
+              // silent fail
+              return of({ annotations: {} as Record<string, Annotation[]> });
+            })
+          )
+      : of(null),
+    // Plus query for forecast data if there is a forecastId stored in the appState.
+    forecastId !== undefined
+      ? (function() {
+          let aggType;
+          const detector = selectedJob.analysis_config.detectors[detectorIndex];
+          const esAgg = mlFunctionToESAggregation(detector.function);
+          if (!modelPlotEnabled && (esAgg === 'sum' || esAgg === 'count')) {
+            aggType = { avg: 'sum', max: 'sum', min: 'sum' };
+          }
+          return mlForecastService.getForecastData(
+            selectedJob,
+            detectorIndex,
+            forecastId,
+            nonBlankEntities,
+            searchBounds.min.valueOf(),
+            searchBounds.max.valueOf(),
+            focusAggregationInterval.expression,
+            aggType
+          );
+        })()
+      : of(null),
+  ]).pipe(
+    map(([metricData, recordsForCriteria, scheduledEventsByBucket, annotations, forecastData]) => {
+      const refreshFocusData = {} as FocusData;
```

Comment:
it appears `focusAnnotationData`, `focusForecastData` and `showForecastCheckbox` should be optional.
is these were made optional, `refreshFocusData` could be built up in one go and avoid the need to assert `FocusData` on an empty object.

Reply:
agree, I kept the original code unchanged here, just combined everything to one callback. will refactor 👌

Reply:
Fixed in https://github.com/elastic/kibana/pull/51008/commits/5c248606b4011d9d5af7e9857706ecbf8f7205d7